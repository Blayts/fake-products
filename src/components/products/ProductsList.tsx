import { useCallback, useEffect, useRef, useState } from 'react';
import type { UIEvent } from 'react';
import { useNavigate, useLocation, useParams } from 'react-router';
import { Alert, List } from 'antd';
import { ProductItem } from './ProductItem';
import { LIMIT_PAGES } from '../../constants/base';
import { useProducts } from '../../hooks/useProducts';
import { debounce } from '../../utils';
import './ProductsList.css';

export function ProductsList() {
    const { category } = useParams();
    const navigate = useNavigate();
    const { pathname } = useLocation();
    
    const [scrollValue, setScrollValue] = useState(1);
    const [page, setPage] = useState(1);
    const { products, loading } = useProducts(category, page);
    const wrapperRef = useRef<HTMLDivElement>(null);

    const changePage = useCallback(
        debounce(() => setPage((page) => page + 1), 500),
        []
    );

    function handleClick(id: number) {
        navigate(pathname + '/product/' + id);
    }

    function handleScroll(e: UIEvent<HTMLDivElement>) {
        if (loading || page >= LIMIT_PAGES) {
            return;
        }

        const div = e.target as HTMLDivElement;
        const toBottom = div.scrollHeight - div.scrollTop - div.scrollWidth;

        if (scrollValue !== toBottom && toBottom < 300) {
            setScrollValue(toBottom);
            changePage();
        }
    }

    useEffect(() => {
        setPage(1);
        wrapperRef.current?.scrollTo?.({ top: 0 });
    }, [pathname]);

    return (
        <div className="products-wrapper" onScroll={handleScroll} ref={ wrapperRef }>
            <List loading={loading}>
                {products.map((product) => (
                    <List.Item
                        key={product.id}
                    >
                        <ProductItem onClick={ () => handleClick(product.id) } product={product}></ProductItem>
                    </List.Item>
                ))}
            </List>
            {page === LIMIT_PAGES && (
                <Alert
                    message="The demo version limit has been reached"
                    showIcon
                    type="warning"
                ></Alert>
            )}
        </div>
    );
}
