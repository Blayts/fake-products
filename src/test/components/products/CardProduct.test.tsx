import * as Router from 'react-router';
import { MemoryRouter } from 'react-router';
import { describe, expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { products } from '../../mock-data/products';
import { CardProduct } from '../../../components/products/CardProduct';

describe('Render card product', () => {    
    test('Snapshot test', async () => {        
        vi.spyOn(Router, 'useParams').mockReturnValue({ id: '1' });
        const component = render(
            <MemoryRouter>
                <CardProduct />
            </MemoryRouter>
        );

        await screen.findByText(products[0].title);

        expect(component.container).toMatchSnapshot();
    })
})