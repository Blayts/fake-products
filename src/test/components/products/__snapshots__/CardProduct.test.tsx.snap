// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render Card Product > Snapshot test 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-7t2xvq"
    style="height: 100%;"
  >
    <div
      class="ant-card-body"
    >
      <div
        class="ant-row css-dev-only-do-not-override-7t2xvq"
      >
        <div
          class="ant-col ant-col-6 css-dev-only-do-not-override-7t2xvq"
        >
          <div
            class="ant-image css-dev-only-do-not-override-7t2xvq"
          >
            <img
              class="ant-image-img css-dev-only-do-not-override-7t2xvq"
              src="Image #1"
            />
            <div
              class="ant-image-mask"
            >
              <div
                class="ant-image-mask-info"
              >
                <span
                  aria-label="eye"
                  class="anticon anticon-eye"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="eye"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                    />
                  </svg>
                </span>
                Preview
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-17 ant-col-offset-1 css-dev-only-do-not-override-7t2xvq"
        >
          <h3>
            Title #1
          </h3>
          <p>
            Descritpion #1
          </p>
          <h4>
            Color: 
            <span
              style="text-transform: capitalize;"
            >
              Color #1
            </span>
          </h4>
          <h4>
            Price: 
            100
            cu
          </h4>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render card product > Snapshot test 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-7t2xvq"
    style="height: 100%;"
  >
    <div
      class="ant-card-body"
    >
      <div
        class="ant-row css-dev-only-do-not-override-7t2xvq"
      >
        <div
          class="ant-col ant-col-6 css-dev-only-do-not-override-7t2xvq"
        >
          <div
            class="ant-image css-dev-only-do-not-override-7t2xvq"
          >
            <img
              class="ant-image-img css-dev-only-do-not-override-7t2xvq"
              src="Image #1"
            />
            <div
              class="ant-image-mask"
            >
              <div
                class="ant-image-mask-info"
              >
                <span
                  aria-label="eye"
                  class="anticon anticon-eye"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="eye"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                    />
                  </svg>
                </span>
                Preview
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-17 ant-col-offset-1 css-dev-only-do-not-override-7t2xvq"
        >
          <h3>
            Title #1
          </h3>
          <p>
            Descritpion #1
          </p>
          <h4>
            Color: 
            <span
              style="text-transform: capitalize;"
            >
              Color #1
            </span>
          </h4>
          <h4>
            Price: 
            100
            cu
          </h4>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test Card Product > Snapshot test 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-7t2xvq"
    style="height: 100%;"
  >
    <div
      class="ant-card-body"
    >
      <div
        class="ant-row css-dev-only-do-not-override-7t2xvq"
      >
        <div
          class="ant-col ant-col-6 css-dev-only-do-not-override-7t2xvq"
        >
          <div
            class="ant-image css-dev-only-do-not-override-7t2xvq"
          >
            <img
              class="ant-image-img css-dev-only-do-not-override-7t2xvq"
              src="Image #1"
            />
            <div
              class="ant-image-mask"
            >
              <div
                class="ant-image-mask-info"
              >
                <span
                  aria-label="eye"
                  class="anticon anticon-eye"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="eye"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                    />
                  </svg>
                </span>
                Preview
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-17 ant-col-offset-1 css-dev-only-do-not-override-7t2xvq"
        >
          <h3>
            Title #1
          </h3>
          <p>
            Descritpion #1
          </p>
          <h4>
            Color: 
            <span
              style="text-transform: capitalize;"
            >
              Color #1
            </span>
          </h4>
          <h4>
            Price: 
            100
            cu
          </h4>
        </div>
      </div>
    </div>
  </div>
</div>
`;
